src/fineman/finemans_algorithm.py

```diff
@@ -0,0 +1,53 @@
+from math import log2
+
+from fineman import preprocess_graph
+from fineman.helper_functions import reweight_graph
+from fineman.dijkstra import dijkstra
+from fineman.elimination_algorithm import elimination_algorithm
+from scripts import generate_double_tree
+
+
+def _inverse_dist_array(source, dists, price_function):
+    actual_dists = [0] * len(dists)
+
+    for i in range(len(actual_dists)):
+        actual_dists[i] = dists[i] + price_function[i] - price_function[source]
+
+    return actual_dists
+
+
+def fineman(graph: dict[int, dict[int, int]], source: int):
+
+    org_graph = graph.copy()
+    n = len(org_graph.keys())
+    m = sum(len(neighbors) for neighbors in org_graph.values())
+
+    graph, neg_edges = preprocess_graph(org_graph, n, m)
+
+    composed_price_function = [0] * len(graph.keys())
+
+    n = len(graph.keys())
+    neg_edges = {(u,v) for u, edges in graph.items() for v, w in edges.items() if w < 0}
+
+    for _ in range(int(log2(n))):
+
+        k = len(neg_edges)
+
+        for _ in range(int(k**(2/3))):
+
+            price_functions = elimination_algorithm(graph, neg_edges)
```

Comment:
This becomes very confusing to read very fast. Again, we should make a decision on when the price functions are applied and make the pattern consistent throughout the code.